torchtune/modules/transformer.py

```diff
@@ -657,18 +765,26 @@ def forward(
         # shape: [b, s, d]
         h = self.tok_embeddings(tokens)
 
-        if self.causal_mask is not None:
+        if self.decoder_caches_are_enabled:
             if mask is not None:
                 raise ValueError(
-                    "An attention mask was set. Cannot use a non-causal mask for inference"
+                    "KV-caches for self-attention layers are setup for inference mode, causal masks must be provided!"
+                    " Use the `mask` arg to provide a causal mask."
+                )
+        if self.encoder_caches_are_enabled:
+            if encoder_mask is not None:
+                raise ValueError(
+                    "KV-caches for cross-attention/fusion layers are setup for inference mode, causal masks must be provided!"
+                    " Use the `encoder_mask` arg to provide a causal mask."
                 )
-            # Track the input position
-            if input_pos is None:
-                input_pos = torch.arange(self.pos, self.pos + seq_len, device=h.device)
-            self.pos = input_pos.max() + 1
-            # shape: [1, input_pos_len, m_s]
-            # in most cases input_pos_len should be 1
-            mask = self.causal_mask[None, input_pos]
+        if (
+            self.encoder_caches_are_enabled
+            or self.decoder_caches_are_enabled
+            and input_pos is None
+        ):
+            raise ValueError(
+                "KV-caches are setup for inference mode, input positions must be provided!"
+            )
```

Comment:
At this point I would offload all of this (plus the seq len validation just above) into `self._validate_inputs` or something. Either way should do this check prior to calling token embeddings though